### PR TITLE
global funding proposal

### PR DESCRIPTION
**Pull request by dobertRowneySr**
Addresses https://github.com/Joystream/joystream/issues/4333

#### Changes
* [x] rename max amount into max total amount (total amount for the funding proposal)
* [x] refactor check in `proposals/codex/lib.rs` to check funding <= max amount
* [x] add failing test for constraint violations
* [x] generate types and metadata

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202132419573087/1203106017956930) by [Unito](https://www.unito.io)

### CODE REVIEW

File: runtime-modules/proposals/codex/src/lib.rs [hunk 1]
```diff
@@ -264,7 +265,7 @@ pub trait Config:
     >;
 
     /// Max amount in funding request proposal (per account)
```
Comment (Lezek123) [inline runtime-modules/proposals/codex/src/lib.rs]:
```suggestion
    /// Maximum total amount in funding request proposal
```

File: runtime-modules/proposals/codex/src/lib.rs [hunk 2]
```diff
@@ -481,8 +482,8 @@ decl_module! {
             ProposalParameters<T::BlockNumber, BalanceOf<T>> = T::UpdateChannelPayoutsProposalParameters::get();
 
         /// Max amount in funding request proposal (per account)
```
Comment (Lezek123) [inline runtime-modules/proposals/codex/src/lib.rs]:
```suggestion
        /// Maximum total amount in funding request proposal
```

File: runtime-modules/proposals/codex/src/lib.rs [hunk 3]
```diff
@@ -750,18 +751,26 @@ impl<T: Config> Module<T> {
...
+                    .try_fold(BalanceOf::<T>::zero(), |sum, el| {
+                        sum.checked_add(&el.amount)
+                    })
+                    .ok_or("Arithmetic Error")?;
+
+                // ensure total_funding_amount in (0, MAX]
+                ensure!(
+                    !total_funding_amount.is_zero(),
+                    Error::<T>::InvalidFundingRequestProposalBalance
+                );
```
Comment (Lezek123) [inline runtime-modules/proposals/codex/src/lib.rs]: I think it's actually worth to check for non-zero amount on each account, not just `total_funding_amount.is_zero()`

File: runtime-modules/proposals/codex/src/lib.rs [hunk 4]
```diff
@@ -50,7 +50,7 @@
     deny(clippy::match_on_vec_items),
     deny(clippy::unreachable)
 )]
-
+#![feature(int_roundings)]
```
Comment (Lezek123) [inline runtime-modules/proposals/codex/src/lib.rs]: This is only required for tests, so how about:
```suggestion
#![cfg_attr(test, feature(int_roundings))]
```

**Review by Lezek123 — CHANGES_REQUESTED** [review]

File: runtime-modules/proposals/codex/src/lib.rs (hunk 3 above)
Comment (ignazio-bovo) [reply runtime-modules/proposals/codex/src/lib.rs:768]: if $x_1, x_2, \dots, x_n$ are nonnegative integers and $x_1 + x_2 + \dots + x_n = 0$ then $x_1 = x_2 = \dots = x_n = 0$

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -1622,7 +1622,7 @@ parameter_types! {
     // Make sure to stay below MAX_BLOCK_SIZE of substrate consensus of ~4MB
     // The new compressed wasm format is much smaller in size ~ 1MB
     pub const RuntimeUpgradeWasmProposalMaxLength: u32 = DispatchableCallCodeMaxLen::get();
-    pub const FundingRequestProposalMaxAmount: Balance = dollars!(10_000);
+    pub const FundingRequestProposalMaxTotalAmount: Balance = 1_000_000 * currency::DOLLARS; // TODO: adjust
```
Comment (Lezek123) [inline runtime/src/lib.rs]: The value should stay as it was:
```suggestion
    pub const FundingRequestProposalMaxAmount: Balance = dollars!(10_000);
```

**Review by Lezek123 — CHANGES_REQUESTED** [review]

**Review by Lezek123 — APPROVED** [review]
LGTM, just needs conflict resolution

**Pull request merged**